File: cslavb/Csla/DataPortal/DataPortalMethodInfo.vb
Imports System.Reflection

Namespace Server

  Friend Class DataPortalMethodInfo

    Private privateRunLocal As Boolean
    Public Property RunLocal() As Boolean
      Get
        Return privateRunLocal
      End Get
      Private Set(ByVal value As Boolean)
        privateRunLocal = value
      End Set
    End Property

    Private privateTransactionalType As TransactionalTypes
    Public Property TransactionalType() As TransactionalTypes
      Get
        Return privateTransactionalType
      End Get
      Private Set(ByVal value As TransactionalTypes)
        privateTransactionalType = value
      End Set
    End Property

    Public Sub New(ByVal info As MethodInfo)
      Me.RunLocal = IsRunLocal(info)
      Me.TransactionalType = GetTransactionalType(info)
    End Sub

    Private Shared Function IsRunLocal(ByVal method As MethodInfo) As Boolean
      Return Attribute.IsDefined(method, GetType(RunLocalAttribute), False)
    End Function

    Private Shared Function IsTransactionalMethod(ByVal method As MethodInfo) As Boolean
      Return Attribute.IsDefined(method, GetType(TransactionalAttribute))
    End Function

    Private Shared Function GetTransactionalType(ByVal method As MethodInfo) As TransactionalTypes
      Dim result As TransactionalTypes
      If IsTransactionalMethod(method) Then
        Dim attrib As TransactionalAttribute = CType(Attribute.GetCustomAttribute(method, GetType(TransactionalAttribute)), TransactionalAttribute)
        result = attrib.TransactionType

      Else
        result = TransactionalTypes.Manual
      End If
      Return result
    End Function

  End Class

End Namespace

File: cslacs/Csla/Server/DataPortalMethodInfo.cs
using System;
using System.Reflection;

namespace Csla.Server
{
  internal class DataPortalMethodInfo
  {
    public bool RunLocal { get; private set; }
    public TransactionalTypes TransactionalType { get; private set; }

    public DataPortalMethodInfo()
    {
      this.RunLocal = false;
      this.TransactionalType = TransactionalTypes.Manual;
    }

    public DataPortalMethodInfo(MethodInfo info)
      : this()
    {
      if (info != null)
      {
        this.RunLocal = IsRunLocal(info);
        this.TransactionalType = GetTransactionalType(info);
      }
    }

    private static bool IsRunLocal(MethodInfo method)
    {
      return Attribute.IsDefined(method, typeof(RunLocalAttribute), false);
    }

    private static bool IsTransactionalMethod(MethodInfo method)
    {
      return Attribute.IsDefined(method, typeof(TransactionalAttribute));
    }

    private static TransactionalTypes GetTransactionalType(MethodInfo method)
    {
      TransactionalTypes result;
      if (IsTransactionalMethod(method))
      {
        TransactionalAttribute attrib =
          (TransactionalAttribute)Attribute.GetCustomAttribute(
          method, typeof(TransactionalAttribute));
        result = attrib.TransactionType;
      }
      else
        result = TransactionalTypes.Manual;
      return result;
    }
  }
}
